driver-core/src/main/com/mongodb/client/model/expressions/BooleanExpression.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+/**
+ * Expresses a boolean value.
+ */
+public interface BooleanExpression extends Expression {
+
+    /**
+     * Returns logical true if this expression evaluates to logical false.
+     * Returns logical false if this expression evaluates to logical true.
+     *
+     * @return True if false; false if true.
+     */
+    BooleanExpression not();
+
+    /**
+     * Returns logical true if this or the other expression evaluates to logical
+     * true. Returns logical false if both evaluate to logical false.
+     *
+     * @param or the other boolean expression.
+     * @return True if either true, false if both false.
+     */
+    BooleanExpression or(BooleanExpression or);
+
+    /**
+     * Returns logical true if both this and the other expression evaluate to
+     * logical true. Returns logical false if either evaluate to logical false.
+     *
+     * @param and the other boolean expression.
+     * @return true if both true, false if either false.
+     */
+    BooleanExpression and(BooleanExpression and);
+
+    /**
+     * If this expression evaluates to logical true, returns the result of the
+     * evaluated left branch expression. If this evaluates to logical false,
+     * returns the result of the evaluated right branch expression.
+     *
+     * @param left the left branch expression
+     * @param right the right branch expression
+     * @return left if true, right if false.
+     * @param <T> The type of the resulting expression.
+     */
+    <T extends Expression> T cond(T left, T right);
```

Comment:
```
one.eq(of(2)).cond(of("yes"), of("no"));
```
The present variant of cond does not supply the original value to the left and right branches. Above, the value is `one`, but it could be a long composite expression. This is inconvenient. (It also fails to imply that the not-taken branch will not be evaluated, since there is no lambda; this problem is also present in `and` and `or`.) 

```
one.condNumber(v -> v.eq(of(2)), v -> of("yes"), v -> of("no"));
```

This alternative has two problems. First, it is always verbose. Second, because there is no self type, distinct variations must be provided for each type, which makes the name confusing (`ifNumber` sounds like a typecheck).

If the value must be supplied, switch can be used instead. The only reason to choose the second variant is to enforce evaluation semantics, in which case the parameters of `and` and `or` should also be lambdas.